File: windows/src/ConnectorControl.Core/State/RestoreModel.cs
using System.Text.Json;

namespace ConnectorControl.Core.State;

/// <summary>Catalog §5 RestoreSheetView state.</summary>
public sealed class RestoreModel : ObservableObject
{
    public const string Headline = "Restore Claude config from a backup";
    public const string Caption = "The current file is backed up first, then replaced by the selected backup.";
    public const string CancelTitle = "Cancel";
    public const string RestoreTitle = "Restore…";
    public const string RestoreButton = "Restore";
    private const string Series = "claude_desktop_config";

    private readonly AppState state;
    private readonly IDialogs dialogs;
    private IReadOnlyList<string> backups = [];
    private string? selection;
    private string? restoreError;

    public RestoreModel(AppState state, IDialogs dialogs)
    {
        this.state = state;
        this.dialogs = dialogs;
    }

    public event Action? CloseRequested;

    /// <summary>Full paths, newest first; the permanent .original snapshot last (catalog §5).</summary>
    public IReadOnlyList<string> Backups => backups;

    public IReadOnlyList<string> BackupNames => backups.Select(Path.GetFileName).Select(n => n ?? "").ToList();

    public string? Selection
    {
        get => selection;
        set
        {
            if (Set(ref selection, value))
            {
                Raise(nameof(CanRestore));
            }
        }
    }

    public bool CanRestore => selection is not null;

    public string? RestoreError
    {
        get => restoreError;
        private set
        {
            if (Set(ref restoreError, value))
            {
                Raise(nameof(HasRestoreError));
            }
        }
    }

    public bool HasRestoreError => restoreError is not null;

    public void Load()
    {
        var found = new List<string>(state.Service.Backups.Backups(Series));
        var original = Path.Combine(state.Service.Backups.BackupsDir, $"{Series}.original.json");
        if (File.Exists(original))
        {
            found.Add(original);
        }
        backups = found;
        Raise(nameof(Backups));
        Raise(nameof(BackupNames));
    }

    public void Cancel() => CloseRequested?.Invoke();

    /// <summary>Confirm, then restore through AppState (which syncs the baseline). True when restored and closing.</summary>
    public bool Restore()
    {
        RestoreError = null;   // a fresh attempt starts with a clean sheet (catalog §5 shows the error only after a failure)
        if (selection is not { } backup)
        {
            return false;
        }
        if (!dialogs.Confirm($"Replace Claude's config with {Path.GetFileName(backup)}?", null, RestoreButton, destructive: true))
        {
            return false;
        }
        try
        {
            state.RestoreClaudeConfig(backup);
            CloseRequested?.Invoke();
            return true;
        }
        catch (Exception ex) when (ex is ClaudeConfigException or IOException or UnauthorizedAccessException or JsonException)
        {
            RestoreError = ex.Message;         // raw message, not Friendly(): catalog §5
            state.LastError = ex.Message;
            return false;
        }
    }
}

File: Sources/ConnectorControlState/RestoreModel.swift
import Foundation
import Combine

/// RestoreSheetView state. The confirmation is a sheet:
/// requestRestore opens it (and clears the previous attempt's error),
/// confirmRestore is its Restore button, cancelRestore its implicit Cancel.
@MainActor
public final class RestoreModel: ObservableObject {
    public static let headline = "Restore Claude config from a backup"
    public static let caption = "The current file is backed up first, then replaced by the selected backup."
    public static let cancelTitle = "Cancel"
    public static let restoreTitle = "Restore…"
    public static let restoreButton = "Restore"
    public static let series = "claude_desktop_config"

    public static func confirmMessage(fileName: String) -> String { "Replace Claude's config with \(fileName)?" }

    private let state: AppState

    /// Full URLs, newest first; the permanent .original snapshot last.
    @Published public private(set) var backups: [URL] = []
    @Published public var selection: URL?
    @Published public private(set) var restoreError: String?
    /// True while the confirmation sheet is up.
    @Published public private(set) var confirming = false

    public init(state: AppState) {
        self.state = state
    }

    public var canRestore: Bool { selection != nil }

    public var confirmMessage: String {
        RestoreModel.confirmMessage(fileName: selection?.lastPathComponent ?? "")
    }

    public func load() {
        do {
            var found = try state.service.backups.backups(series: RestoreModel.series)
            if let original = state.service.backups.originalSnapshotURL(series: RestoreModel.series) {
                found.append(original)
            }
            backups = found
        } catch {
            restoreError = error.localizedDescription
        }
    }

    /// The Restore… button. A fresh attempt starts with a clean sheet: the
    /// previous attempt's error must not outlive a new selection or a cancelled
    /// confirmation. Then the confirmation sheet opens.
    public func requestRestore() {
        restoreError = nil
        guard selection != nil else { return }
        confirming = true
    }

    public func cancelRestore() { confirming = false }

    /// The sheet's Restore button: restore through AppState (which syncs the
    /// baseline). True when restored and the sheet should close.
    public func confirmRestore() -> Bool {
        confirming = false
        guard let backup = selection else { return false }
        do {
            try state.restoreClaudeConfig(from: backup)
            return true
        } catch {
            restoreError = error.localizedDescription   // raw message, not friendly()
            state.lastError = error.localizedDescription
            return false
        }
    }
}
